conv2d

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* #61188
* #61150
* **#61093**

Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

migeed-z committed:
conv2d

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562478).

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
+        return Dyn
+
+    elif isinstance(h_in, int):
+        h_out = floor((h_in + (2 * op_type.padding[0] - op_type.dilation[0] *
```
nit: i would factor out `calculate_hout` and `calculate_wout`, and parameterize by name/index, especially since Conv3d is the same exact logic for the fifth dimension.

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
+    n.args[0].type = apply_matching(n.args[0].type, 4)
+    arg_type = n.args[0].type
+    n.type = apply_matching(n.type, 4)
+    if is_consistent(arg_type.__args__[1], op_type.in_channels) and \
```
nit: i dont know that you actually need the `\` here...

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
+    assert isinstance(n.args[0], Node)
+    n.args[0].type = apply_matching(n.args[0].type, 4)
+    arg_type = n.args[0].type
+    n.type = apply_matching(n.type, 4)
```
what does this line accomplish ? since we re-assign to n.type below

migeed-z replied:
this is for checking if the current node can be expanded to a size 4 tensor so we can check for consistency in the following conditional. I can do this directly in the conditional though instead!

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
+        return Dyn
+
+    elif isinstance(w_in, int):
+        w_out = floor((w_in + (2 * op_type.padding[1] - op_type.dilation[1] *
```
nit: I would rewrite `floor(...)` to just use floor division `(op_type.kernel_size[1] - 1) - 1)) /` -> `(op_type.kernel_size[1] - 1) - 1)) //`

it's probably unlikely that floor(`x/y`) is different than `x // y` but sometimes floating point math is weird like that lol, and i think doing it this way is more in line with how it is computed in eager..

eellison reviewed (commented):
lgtm!  just small comments

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -207,6 +209,53 @@ def bn2d_inference_rule(n: Node, module_instance):
+        raise TypeError(f'{w_in} in {op_type} must be a number or Dyn')
+
+@register_inference_rule(Conv2d)
+def conv2d_inference_rule(n: Node, op_type):
```
Similarly to the previous PR, can you rename `op_type` and add some docs?

jamesr66a commented on test/fx/test_gradual_type.py:
```diff
@@ -7,6 +7,11 @@
+def conv3x3(in_planes, out_planes, stride=1, groups=1, dilation=1):
+    """3x3 convolution with padding"""
+    return torch.nn.Conv2d(in_planes, out_planes, kernel_size=3, stride=stride,
+                           padding=dilation, groups=groups, bias=False, dilation=dilation)
```
is `padding=dilation` intended?

migeed-z replied:
yeah, I copied this function from torchvision_models.py and it has padding = dilation for some reason.

jamesr66a commented on test/fx/test_gradual_type.py:
```diff
@@ -330,6 +335,76 @@ def forward(self, x: Dyn):
         with self.assertRaises(TypeError):
             tc.type_check()
 
+    def test_type_check_conv2D(self):
```
I think a couple things would also make the tests more robust:

1) Can we test that the inferred sizes match with the sizes we actually see when running the module?
2) Can we make the testing more exhaustive? e.g. automate testing over different `kernel_size`, `stride`, `padding`, `dilation`, `groups` parameters and make sure it works across all those parameters

This PR might take a while to get fully tested/correct so it might be a good idea to pull it out from the stack and develop it separately

jamesr66a reviewed (commented):
Cool, as mentioned inline, there are a lot of degrees of freedom in convolution here and it would be good to try to more thoroughly test across those

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z committed:
Update on "conv2d"


Differential Revision: [D29562478](https://our.internmc.facebook.com/intern/diff/D29562478)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562478).

jamesr66a reviewed (approved):
![image](https://user-images.githubusercontent.com/4685384/125005868-ad361280-e011-11eb-9a0e-6eaf0f65a1ac.png)

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -207,6 +208,65 @@ def bn2d_inference_rule(n: Node, module_instance):
     else:
         raise TypeError(f'Cannot apply {module_instance} with input type {arg_type} and existing type {n.type} on {n}')
 
+def calculate(d_in, module_instance, index):
```
Can this be named something more descriptive?
